File: Programmers/JeongHyun/3주차(Stack, Queue)/프린터.js
function solution(priorities, location) {
    var answer = 0;
    //arr = [priorities[i] , i] 로 value/index 를 저장해주기 위한 변수
    let arr = priorities.map((value,index)=>value = [value,index]);
    /*
    1. 우선 순위 최대값 찾기
    2. 대기목록에서 J보다 중요도가 높은 문서가 한개도 없을 때까지 pop하고 push해서 우선 순위가 가장 높은 것이 앞으로 오게 만들기
    3. J를 인쇄하기(pop)해주기(answer++)  => location과 맞는 것이 나올때까지 반복 (x라면 1번으로) (o라면 4번으로)
    4. 같은 것이 나오면 answer값 리턴 
    */
    while(1){ 
        //최대값 찾기
        let maxValue = -1;
        for(let i=0;i<arr.length;++i){
            if(maxValue<arr[i][0])maxValue =arr[i][0];
        }
        //대기목록에서 J보다 중요도가 높은 문서가 한개도 없을 때까지 pop하고 push해서 우선 순위가 가장 높은 것이 앞으로 오게 만들기
        while(maxValue!=arr[0][0]){
            arr.push([arr[0][0],arr[0][1]]);
            arr.splice(0,1);
        }
        //location과 맞는 것인지 여부판단
        if(arr[0][1]==location) return answer+=1;
        else{
            //아니라면 우선순위가 가장 높은 것 인쇄하고 재정렬
            arr.splice(0,1);
            answer++;
        }
    }
    return answer;
}

File: Programmers/SeungDo/3주차(Stack, Queue)/프린터.cpp
#include <string>
#include <vector>

using namespace std;
// 처음에 우선순위가 가장 높은 값을 구한 뒤 벡터의 처음값이 max와 같고 location이 0일때 리턴
int solution(vector<int> priorities, int location) {
    int answer = 0 ,max = 0;
    for(int i = 0; i<priorities.size();i++) if(max < priorities[i]) max= priorities[i];
    while(1){   
        if(priorities[0] == max) {
            if(location == 0) return answer + 1; // 사용자가 원하는 작업일때 리턴
            priorities.erase(priorities.begin()); // 팝
            max =0;
            for(int i = 0; i<priorities.size();i++) if(max < priorities[i]) max= priorities[i];
            answer++; // 작업 수 ++
        }
        else{
            priorities.push_back(priorities[0]); // 첫 번째 값을 맨뒤에 푸쉬
            priorities.erase(priorities.begin()); //첫 번째 값 팝 
        }
        location--; 
        if(location == -1)location = priorities.size() - 1;
    }
}
